### PR TITLE
[CORDA-303]: Add some tests that stop and start a node and check state is persistent.

### PR DESCRIPTION
**Pull request by sollecitom**
https://r3-cev.atlassian.net/browse/CORDA-303

### CODE REVIEW

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 1]
```diff
@@ -191,7 +191,15 @@ sealed class NodeHandle {
             override val webAddress: NetworkHostAndPort,
             val debugPort: Int?,
             val process: Process
-    ) : NodeHandle()
+    ) : NodeHandle() {
+
```
Comment (fenryka) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: remove blank line

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 2]
```diff
@@ -191,7 +191,15 @@ sealed class NodeHandle {
...
+        override fun stop(): CordaFuture<Unit> {
+
```
Comment (fenryka) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: and here

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 3]
```diff
@@ -200,9 +208,23 @@ sealed class NodeHandle {
             override val webAddress: NetworkHostAndPort,
             val node: Node,
             val nodeThread: Thread
-    ) : NodeHandle()
+    ) : NodeHandle() {
+
+        override fun stop(): CordaFuture<Unit> {
+
```
Comment (fenryka) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: and here

File: node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                val page = it.proxy.vaultQuery(MessageState::class.java)
+                val retrievedMessage = page.states.singleOrNull()?.state?.data?.message
+                assertEquals(message, retrievedMessage)
+            }
+        }
+    }
+}
+
+fun isQuasarAgentSpecified(): Boolean {
+
```
Comment (fenryka) [inline node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt]: blank line

File: node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    val jvmArgs = ManagementFactory.getRuntimeMXBean().inputArguments
+    return jvmArgs.any { it.startsWith("-javaagent:") && it.endsWith("quasar.jar") }
+}
+
+@CordaSerializable
+data class Message(val value: String)
+
+data class MessageState(val message: Message, val by: Party, override val linearId: UniqueIdentifier = UniqueIdentifier()) : LinearState, QueryableState {
```
Comment (fenryka) [inline node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt]: I'd wrap this line, it's kinda long but up to you :)

File: node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+    override val contract = MessageContract()
+    override val participants: List<AbstractParty> = listOf(by)
+
+    override fun generateMappedObject(schema: MappedSchema): PersistentState {
+
```
Comment (fenryka) [inline node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt]: blank line

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 7]
```diff
@@ -200,9 +208,23 @@ sealed class NodeHandle {
...
+            node.stop()
+            nodeThread.interrupt()
```
Comment (fenryka) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: could combine these with a with to avoud
X.something
X,somethingelse

so with(x) {
something
somethingelse
}

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 8]
```diff
@@ -191,7 +191,15 @@ sealed class NodeHandle {
             override val webAddress: NetworkHostAndPort,
             val debugPort: Int?,
             val process: Process
-    ) : NodeHandle()
+    ) : NodeHandle() {
+
+        override fun stop(): CordaFuture<Unit> {
+
+            process.destroy()
```
Comment (fenryka) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: combine with with or apply

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt (hunk 7 above)
Comment (sollecitom) [reply testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Good point, fixed it here and in the OutOfProcess override.

File: node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+import kotlin.test.assertEquals
+
+class NodeStatePersistenceTests {
+
+    @Test
+    fun `persistent state survives node restart`() {
+        val user = User("mark", "dadada", setOf(FlowPermissions.startFlowPermission<SendMessageFlow>()))
+        val message = Message("Hello world!")
+        driver(isDebug = true, startNodesInProcess = isQuasarAgentSpecified()) {
+
```
Comment (fenryka) [inline node/src/integration-test/kotlin/net/corda/test/node/NodeStatePersistenceTests.kt]: blank line

**Comment by fenryka** [comment]
remember to squash commits before push

**Pull request merged**